components/eam/src/physics/crm/crm_rad_module.F90

```diff
@@ -8,65 +8,91 @@ module crm_rad_module
+      real(crm_rknd), allocatable :: ns(:,:,:,:) ! rad cloud snow crystal number (#/kg)
    end type crm_rad_type
 
 contains
 
    !------------------------------------------------------------------------------------------------
    ! Type-bound procedures for crm_rad_type
-   subroutine crm_rad_initialize(rad)
+   subroutine crm_rad_initialize(rad, ncrms, crm_nx_rad, crm_ny_rad, crm_nz)
       class(crm_rad_type), intent(inout) :: rad
+      integer, intent(in) :: ncrms, crm_nx_rad, crm_ny_rad, crm_nz
 
-      ! Nullify pointers
-      rad%qrad => null()
-      rad%temperature => null()
-      rad%qv => null()
-      rad%qi => null()
-      rad%cld => null()
+      if (.not. allocated(rad%qrad))        allocate(rad%qrad       (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%temperature)) allocate(rad%temperature(ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%qv))          allocate(rad%qv         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%qc))          allocate(rad%qc         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%qi))          allocate(rad%qi         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%cld))         allocate(rad%cld        (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      
+      if (.not. allocated(rad%nc))          allocate(rad%nc         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%ni))          allocate(rad%ni         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%qs))          allocate(rad%qs         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%ns))          allocate(rad%ns         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
 
-      rad%nc => null()
-      rad%ni => null()
-      rad%qs => null()
-      rad%ns => null()
+      call prefetch(rad%qrad)
+      call prefetch(rad%temperature)
+      call prefetch(rad%qv)
+      call prefetch(rad%qc)
+      call prefetch(rad%qi)
+      call prefetch(rad%cld)
+
+      call prefetch(rad%nc)
+      call prefetch(rad%ni)
+      call prefetch(rad%qs)
+      call prefetch(rad%ns)
+
+      rad%qrad           = 0
+      rad%temperature    = 0
+      rad%qv             = 0
+      rad%qc             = 0
+      rad%qi             = 0
+      rad%cld            = 0
+
+      rad%nc             = 0
+      rad%ni             = 0
+      rad%qs             = 0
+      rad%ns             = 0
 
    end subroutine crm_rad_initialize
    !------------------------------------------------------------------------------------------------
    subroutine crm_rad_finalize(rad)
       class(crm_rad_type), intent(inout) :: rad
 
-      ! Nullify pointers
-      rad%qrad => null()
-      rad%temperature => null()
-      rad%qv => null()
-      rad%qi => null()
-      rad%cld => null()
-
-      rad%nc => null()
-      rad%ni => null()
-      rad%qs => null()
-      rad%ns => null()
+      deallocate(rad%qrad)
```

Comment:
It might be wise to add `if (allocated(rad%qrad)) deallocate(rad%qrad)` patterns here for safety, but probably doesn't matter.

Reply:
We didn't do that in the other crm type modules, but you're right, I'll add it.

Reply:
done.